Phase 2.B - Feature functional component

carlospgraciano opened the pull request:
**Details**

This pull request contains the addition of the MantraForm functional component that handles the rendering of components in Mantra. The following actions were taken in order to add the MantraForm functional component:

- **Created** a **functional component** to be the MantraForm component that interacts with metadata of the Router instance.
- **Created** route **utilities** **to** **convert** the Route params into a path and **to append** information to the Route params.
- **Created** the **MantraForm class**--inherits from MantraComponent class--that contains instance and static methods to generate a MantraData object containing enough information to render a component.
- **Created** the **MantraData clas**s that contains all necessary information to initiate Mantra and render a component in the functional component.
- **Modify** the **MantraForm mixin** to add the "$m_mixin" to indicate that the mixin type is "MantraForm".
- **Integrated** the MantraForm configuration with the MantraForm functional component also used the route utilities.

Each class and/or function created or modified in this pull request was documented using the JSDoc convention.

PD: Some commits may show no changes at all or some have the name of GitHub that is because I had issues with Git. Git in both the GitKraken and Terminal was tracking the files inaccurately which produced these "ghost" commits that reflected no changes at all. Nothing to worry because no changes were lost or applied.

Commit:
Added a getter for the store schemas module

Commit:
New core component folder for configuration classes

A Base MantraComponent class was created for MantraForm class to use; Added documentation in MantraForm class; Added a new getConfig static method to create a MantraData

Commit:
Move the MantraForm class location to the core folder

Commit:
Change how the MantraMixin access the stores schema module

Commit:
Added a new Mantra Functional Component

Commit:
Added the MantraData class and its documentation

carlospgraciano committed:
MantraForm getConfig flow modifications

The following changes were added to the MantraForm:

- Full sequence to obtain the component configuration based on MantraData object.

- Remove the path parameter in the setBase method because the MantraData sets the path as a property in its instantiation.

carlospgraciano committed:
New setRecord flow and modification in setBase flow

The following changes were applied to the MantraData class:

- Declaration of the baseComponent property in the class constructor.

- Documentation of the responsibilities of each class method.

- Added the setRecord flow and the methods the flow needs (setId, setRelationship, isId, ...).

- Added the update and trim path methods.

carlospgraciano committed:
Upload of file modifications

Certain files were detected as modified by git, therefore, this commit is to update the GitHub repository with this changes.

Commit:
GitHub detection of files modifications

carlospgraciano committed:
Added MantraData's base class property

The following changes were made in order to add the "baseClass" property:

- New constant containing all of the supported MantraComponent classes.
- New parameter in MantraData's constructor
- New MantraData's method to get the base class property value.
- New MantraData method to set the base class property using the get base class method.
- Use of the base class set method in MantraData's constructor.
- Documentation for each new method and the new constant.

carlospgraciano committed:
Rename to config the MantraData instantiation in MantraForm class

Previously, MantraData was the name of the constant that store the object returned by the instantiation of the MantraData inside the MantraForm class. But there was a naming conflict because two parts of the MantraForm file declares the same name: 1. Import statement (which is a global constant) and 2. getConfig method (which is a local constant). The output result was an error because there can't be a global and local constants with the same name.

Currently, config is the name of the constant inside the getConfig method which don't produces the naming conflict inside the MantraForm file.

carlospgraciano committed:
Set the action property of MantraData class

The action property is setted using the _path, schema and id properties of the MantraData class, as well as, the DEFAULT_ACTIONS constant.

To achieve the set of the action property the following changes were made:

- Created the "DEFAULT_ACTIONS" constant that contains Mantra's built-in actions such as "create" and "update" actions.

- Created the "setAction" method.

- Created the "getAction" method that--based on the _path property--retrieves the corresponding action for MantraData.

- Created the "isCustomAction" method that verifies if the first string in the "_path" property is a custom action (using the schema property).

- Created the "getCustomAction" method that retrieves--from the schema property--the custom action that matches the first string in the "_path" property.

- Created the "isUpdateAction" method that verifies if the default action should be the "update" action based on the ID having an "empty" value.

- Created the "getDefaultAction" method that retrieves--from the "DEFAULT_ACTIONS" constant--the default action that the MantraData requires.

Extra work:

- Missing semicolons were added to MantraData's methods.

- The documentation of the "getBaseClass" method was modified to be more descriptive.

- The order in the body of the "setBase" method was modified in order to reuse the "name" constant in other methods.

- The assignment--in the constructor--of the "baseClass" property was modified to remove unnecessary code.

carlospgraciano committed:
Set the component property of MantraData class

The component property of a MantraData class is set using static methods from the _baseClass property. The _baseClass is a component class that contains static methods to retrieve a component that the MantraData needs to set.

To achieve the set of the component property, the following changes were made:

In the MantraForm class:

- Added new imports of get and cloneDeep methods of lodash.

- Created a MIXIN constant to verify if a MantraComponent object contains the valid MantraForm mixin.

- Created the getComponent static method to retrieve a MantraForm component from a schema.

- Created the getPath static method to generate a path to traverse a schema.

- In the getConfig method, added a try/catch statement to handle the errors that may be produce by any of the MantraData's set methods.

In the MantraData class:

- Added the setComponent method that set the component property of the MantraData class.

- In the setBase method, added a throw error in case. no schema was found.

In the MantraComponent class:

- Removed the isValidComponent method from the class and let each MantraComponent implement it as a static method.

In the package json:

- Added lodash as a dependency to be used by the MantraForm class.

carlospgraciano committed:
Set the endpoint property of MantraData class

In order to achieve the set of the endpoint property, the following changes were made:

In the MantraData class:

- Created the "endpoint" property and set it to it's default value which is a string.

- Created the "getEndpoint" method which uses a regex to replace all dots inside the "path" property to be a slash and then returns the replaced string.

- Created the "setEndpoint" method which retrieves the endpoint value from the "getEndpoint" method and then sets the endpoint property to the endpoint value.

- Added documentation for the set and get methods of the endpoint property.

- Modify the documentation of the trimPath, getCustomAction, and getDefaultAction to be more descriptive regarding their returned value.

In the MantraForm class:

- Called the "setEndpoint" method inside the "getConfig" static method.

carlospgraciano committed:
Move the isValidComponent to MantraComponent class

Previously, only the MantraForm had the isValidComponent method which wouldn't be reusable among all the classes that inherit from the MantraComponent class.

Since the ultimate goal is that every MantraComponent class has the implementation of this method then I decided to move it from the MantraForm to the MantraComponent.

carlospgraciano committed:
Added mixin type property for mixin validations

I added the "$m_MantraForm" property inside the mixin to validate if the mixin is being used by a MantraForm vue component.

In the MantraComponent class:
- Changed the property of the component to point to the mixin type property in order to validate using the correct information.

carlospgraciano committed:
Modify the isValidComponent parameter description

Commit:
Changes due to testing of MantraDaa creation and functional component rendering

carlospgraciano committed:
WIP - Integrating MantraData and MantraForm in functional component

Commit:
WIP - Integrating MantraData and MantraForm in functional component

carlospgraciano committed:
Rewrote the _getPath logic and removed unused code from getComponent

carlospgraciano committed:
WIP - Rendering of the component based on the config object

carlospgraciano committed:
Re-wrote the component validation method

The new component validation method was created with the purpose of be reused among all the MantraComponent classes since all the information it uses it's based on inheritance and vue instance.

carlospgraciano committed:
Change the mixin type property name

carlospgraciano committed:
Added the context property to MantraData

In this changes, the context property was only added in the constructor. The setters and getters of the context property will be added in future commits.

carlospgraciano committed:
Added documentation and code organization

jonatangerman commented on lib/components/MantraForm.vue:
```diff
@@ -0,0 +1,39 @@
+<script>
+import MantraForm from './../core/components/MantraForm';
+
+const convertRouteToPath = function (route) {
```
Rename method into something more 'generic' like for example 'getPathFromObject' as stated in the documentation, or another naming similar in the 'coverage' of the method transform, since is not exclusively for routes but for essentially any object.

carlospgraciano replied:
Sure! Will do.

jonatangerman commented on lib/components/MantraForm.vue:
```diff
@@ -0,0 +1,39 @@
+    beforeRouteEnter(to, from, next) {
+        const path = convertRouteToPath(to);
+        const config = MantraForm._getConfig({ path });
+        const IS_PATH_EMPTY = (config._path === '');
```
[New suggestion] Insert this IS_PATH_EMPTY evaluation as part of the MantraForm._getConfig() and throw an error from in reflecting the path was not completely resolved.
Try catch the conditioning for the route, so the try blocks resolves in the first evaluation of the current if statement and the catch console.error() the error message and resolves with the current else statement (Gotta add suggestion to the documentation)

carlospgraciano replied:
Sure! Will do.

jonatangerman commented on lib/components/MantraForm.vue:
```diff
@@ -0,0 +1,39 @@
+    render(createElement, context) {
+        const { props: { config }, parent, children, data } = context;
+
+        if (!config.component._isValidComponent(parent)) 
```
Awesome Approach! (Even more since the documentation didn't contemplate it)
I found the 404 component could be a little misleading since a wider range of not only the component being found could be the trigger of rendering a 404 component, but! Those are things that we could refine when the framework is at use and as user feedback! This conditional here is an excellent approach.
Not sure if this could be a 'try/catch scenario' but again, is something that we could refine on release / user feedback.

In the time being I could only suggest that for consistency the if / else statement could be reduced to a single if statement with a return and avoid the else statement making it see just as the 'default return' of a function. This is even one of the suggested rules for the AirBnB linter, that we could use as guide in our coding. :) https://eslint.org/docs/rules/no-else-return

Also lets be consistent and in all if statements lets use { }

carlospgraciano replied:
Sure! Will do.

Also, thanks for the suggestion regarding the "no-else-return" 'cause I didn't know about it.

jonatangerman commented on lib/core/MantraData.js:
```diff
@@ -0,0 +1,303 @@
+/*
+    This function consider a number any value that is not null or NaN.
+*/
+const IS_ID_NUMBER = (value) => {
```
We could skip the _NUMBER part in the naming, since the essential thing we are validating is the 'definition of ID in our system' being a number is just the 'current evaluation' but for example if our system scales to support uuid, the IS_ID_NUMBER() naming could be misleading and IS_ID() could be more inclusive

Also our boolean constant lets keep them being a UPPERCASE/SNAKE naming.

carlospgraciano replied:
Following the suggestion made in the _isId method then this ÏS_ID_NUMBER method will be removed and the content of the function's body will be added in the _isId method.

jonatangerman commented on lib/core/MantraData.js:
```diff
@@ -0,0 +1,303 @@
+    _isSchema(name) {
+        const schemas = MantraPlugin.getSchemasModule();
+
+        if (schemas.hasOwnProperty(name)) 
```
Return the evaluation of the hasOwnProperty(name) since the boolean itself is the returning value so the if/else block statement is unneeded ;)
We could also apply this rule https://eslint.org/docs/rules/no-prototype-builtins in order to uses of 'hasOwnProperty' evaluations since it is more canonical and safe at the moment of evaluating prototypes, and we might get to dig into that in a near future ;)

carlospgraciano replied:
Ohhh! Wao! It makes a lot of sense. I didn't think of using the static method from the Object class. Thank you for showing me this approach and I will apply the changes right away.

jonatangerman commented on lib/core/MantraData.js:
```diff
@@ -0,0 +1,303 @@
+    _isRelationship(name) {
+        const relationships = this.schema.relationships;
+        const HAS_RELATIONSHIPS = (relationships !== undefined);
+        return HAS_RELATIONSHIPS && relationships.some((rel) => rel.name === name);;
```
Make the evaluation after the '&&' a constant as well. Lets keep consistent the intention of our boolean constant when we declare them, we either declare them all, or none :)

carlospgraciano replied:
I tried to create a constant for the "some" evaluation but I ineffectively couldn't achieve it. 

The problem is that the "some" method expected the "relationships" to be always an array and if not it will throw an exception.

It doesn't mean I can't apply the changes but we need to be aware that the error can happen. Also, since we have try/catch statements in the "_getConfig" method of the MantraForm class then these errors will be resolved by a console.error().

The code modification will look like:
```suggestion
        const IS_RELATIONSHIP = relationships.some((rel) => rel.name === name);
        return HAS_RELATIONSHIPS && IS_RELATIONSHIP;
```

jonatangerman replied:
You are totally right, this cannot be achieved with different constants. My bad, didn't saw it that way. We might need another techniche to achieve this, such as 'fast exits' for example if the property is already undefined return the method for example.

But that syntax will be more confusing In my opinion, so we better stick to the boolean evaluation of javascript (your first approach) making the evaluation in the same line so the boolean evaluation do not crash if relationship for example is undefined.

Disregard my suggestion and pls leave the code at first pull requested. Thx Mate!

jonatangerman commented on lib/core/MantraData.js:
```diff
@@ -0,0 +1,303 @@
+     * @returns {boolean} - True if the relationship name exists in the schema property.
+     */
+    _isId(value) {
+        return IS_ID_NUMBER(value);
```
Since we are not importing/exporting as an utility the IS_ID_NUMBER() method and is only referenced through this private method _isId() we might, get rid of the IS_ID_NUMBER() method and put its current body into the _isId(). I guess is like the get/set property dilemma, but with utilities we could use this suggested approach meanwhile, then in the future we could reevaluate it.

carlospgraciano replied:
Definitely, I think I was overdoing the implementation of the _isId method by thinking ahead by creating a method to check if an ID is a number. But as of right now, our only check is if its a number and therefore the body of the _isId() should contain that verification.

Will apply changes.

jonatangerman commented on lib/core/MantraData.js:
```diff
@@ -0,0 +1,303 @@
+     * Gets a custom action or a default action object based on the _path property.
+     * @returns {object} - A default or custom action object.
+     */
+    _getAction() {
```
Implement the https://eslint.org/docs/rules/no-else-return rule

carlospgraciano replied:
Sure! Will do.

jonatangerman commented on lib/core/MantraData.js:
```diff
@@ -0,0 +1,303 @@
+    _isCustomAction() {
+        const [action = ''] = this._path.split('.');
+        const HAS_ACTIONS = this.schema.actions;
+        return HAS_ACTIONS && this.schema.actions.hasOwnProperty(action);
```
Create every needed Boolean evaluation as constant

carlospgraciano replied:
This case is similar to the one in the "_isRelationship" method because if I access the "hasOwnProperty" when "actions" is not a valid object then an error will be throw in the application.

It doesn't mean I can't apply the changes but we need to be aware that the error can happen. Also, since we have try/catch statements in the "_getConfig" method of the MantraForm class then these errors will be resolved by a console.error().

The code modification will look like:
```suggestion
        const IS_ACTION = Object.prototype.hasOwnProperty.call(this.schema.actions, action);
        return HAS_ACTIONS && IS_ACTION;
```

jonatangerman replied:
Yeah bro my bad, let's leave the javascript boolean evaluation as it is, it is clearer! Thanks!

jonatangerman commented on lib/core/MantraData.js:
```diff
@@ -0,0 +1,303 @@
+     * @returns {object} - A built-in Mantra action object.
+     */
+    _getDefaultAction() {
+        if (this._isUpdateAction()) 
```
Implement the https://eslint.org/docs/rules/no-else-return rule

carlospgraciano replied:
Sure! Will do.

jonatangerman commented on lib/core/MantraData.js:
```diff
@@ -0,0 +1,303 @@
+     */
+    setComponent() {
+        const component = this._baseClass.getComponent(this);
+        this.component = component;
```
Let be consistent with the way we end up assigning values in void methods that its final goal is to modify a class property. There are current two alternatives being implemented at this file.
1) Create a const with the evaluation and assign the const to the property
2) Assign the property directly to the evaluation (skip the const creation) (for example at the _updatePath() method)

I suggest we take only one approach for this scenarios and I would suggest it to be the option 2).

carlospgraciano replied:
Sure! Will do the option 2.

jonatangerman commented on lib/core/MantraData.js:
```diff
@@ -0,0 +1,303 @@
+     */
+    setEndpoint() {
+        const endpoint = this.getEndpoint();
+        this.endpoint = endpoint;
```
Same suggestion as with the one at the setComponent() method

carlospgraciano replied:
Sure! Will do.

jonatangerman commented on lib/core/MantraData.js:
```diff
@@ -0,0 +1,303 @@
+     * Gets a endpoint string action object based on MantraData's properties.
+     * @returns {string} - A string concatenated by slash.
+     */
+    getEndpoint() {
```
Same suggestion as with the one at the setComponent() method

carlospgraciano replied:
Sure! Will do.

jonatangerman commented on lib/core/components/MantraForm.js:
```diff
@@ -0,0 +1,63 @@
+     * @returns {MantraData} - The MantraForm component config.
+     */
+    static _getConfig({ path }) {
+        const config = new MantraData({ path, baseClass: MantraForm });
```
We decided to keep the classes used as Strategy (for now) to be resolved in the target class (in this case MantraData) with closures that avoid the circular reference to be triggered at compilation time.

This translate into the baseClass param to be just a String 'MantraForm' and the actual resolving of the MantraForm as class occurs at the MantraData class using its getter baseClasses instead of the module constant

carlospgraciano replied:
Yes! Thank you for reminding me 'cause I forgot this is a "must do" change.

Will apply the changes right away

jonatangerman commented on lib/core/components/MantraForm.js:
```diff
@@ -0,0 +1,63 @@
+     */
+    static _getPath(data) {
+        const pathInfo = ['actions', data.action.name, data.context];
+        const filteredInfo = pathInfo.filter((str) => str);
```
What is the purpose of this filter()? Since the const pathInfo is being set manually with the properties to handle and join(). I see no point in a filter method, could you elaborate pls.

carlospgraciano replied:
This is a temporary filter and it's due to the fact that the "data.context" is not yet implemented. The "data.context" property is set to be an empty string in the constructor of the MantraData class. If you join the following string:

`const pathInfo = ['actions', 'update', ''];`
`const path = pathInfo.join('.'); // Returns 'actions.update.'`

That path will be invalid since the "lodash/get" method will not be able to find any information within the "schema" property. 

So, to avoid it from happening I filtered the array because I wanted to do the least modification possible to the "_getPath" method. My plan was that once the "context" property is fully implemented then that filter will be removed.

jonatangerman replied:
Awesome! Excellent approach mate! Disregard my review on this part.

jonatangerman commented on lib/core/data-types/Factory.js:
```diff
@@ -2,6 +2,9 @@ import MantraDataType from './MantraDataType';
+/*
+    This object contains all the supported MantraDataTypes.
+*/
 const DATA_TYPES = {
```
Let's take this opportunity to move the DATA_TYPES const serving as 'core of our factory' to the create() method and pass it by params to the method needed it. Or even it could be a get () on the class (your pick ;))

This is in order to avoid circular dependencies by moving these constant from being a module property and let then be evaluated in a 'non-sync' execution for the module resolving such as closured in a method or obtained by a class getter.

carlospgraciano replied:
Sure! Will do.

I will pick the "method" approach to be consistency because the MantraData already uses a method to obtain all of the MantraComponent classes in order to avoid any circular dependencies.

jonatangerman commented on lib/core/data-types/Factory.js:
```diff
@@ -34,7 +37,7 @@ export default class DataTypeFactory {
-     * @return {MantraDataType} A mantra data type object.
+     * @returns {MantraDataType} - A mantra data type object.
      */
     static create(params) {
```
Remember to update the methods using the DATA_TYPES constant to the new way to declare it outside of the module scope.

carlospgraciano replied:
Sure! Will do.

Commit:
Change convertRoutePath to getPathFromObject

The name of the convertRouteToPath method was changed to the generic getPathFromObject because it described better that it takes objects and returns a path based on them.

Commit:
Change the name of the path method & Added try/catch to the router guard

Commit:
Removed the try/catch from getConfig and Added the IS_PATH_EMPTY evaluation

Previously, the getConfig try/catch would handle any error from the getConfig. Currently, the getConfig don't handle any error catching, instead the functional component will handle the error catching.

Commit:
Change the if statements to follow the no-else-return rule from Airbnb style guide

Commit:
Removed the IS_ID_NUMBER utility and added the content of its body to the isId method

Commit:
Change the isSchema method to use the approach of Airbnb style guide called no-prototype-builtins

Commit:
Removed unnecessary constant and directly assing the getters value to component and endpoint properties

carlospgraciano commented on lib/components/MantraForm.vue:
```diff
@@ -0,0 +1,42 @@
+
+        try {
+            const config = MantraForm._getConfig({ path });
+            console.log("CONFIG", config);
```
Note: This will be removed after all the suggestions are applied.

jonatangerman replied:
Got it

Commit:
Removed the DATA_TYPES constant from the Factory module to a separate module

Previously, the Factory module had a constant reflecting all of the supported Mantra data type classes but this could cause circular dependencies if any of data type classes dependend on the Factory module. Currently, the DATA_TYPES constant is in the ./index module of the DataTypes folder and exports as a default a function that returns all the supported data types because functions don't produce circular dependencies.

Commit:
Change the baseClass constructor param to be a string

Previously, the MantraData module was passed through a param the definition of the baseClass it will need to store as a property.  Currently, instead of a class definition, a string with the name of the class is passed as the baseClass parameter. This string will be used to search for the class definition and set it in the _baseClass property of MantraData. To avoid circular dependencies, the MantraData imports the getComponentClasses function to retrieve all supported MantraComponent classes in an object that we call BASE_CLASSES and from that object we select the baseClass the MantraData needs to store.

Commit:
Removed an unnecessary console log from the functional component

Commit:
Change the constant scope from local (in a function) to global (in a module)

Commit:
Change the BASE_CLASSES constant scope to the _getBaseClass local scope

Commit:
Change the DATA_TYPES constant scope to the getClass and isValidType local scope

Commit:
Change the IS_PATH_EMPTY to isValid and add it to the MantraData class

Commit:
Added a getter for the BASE_CLASSES in the MantraData class

Commit:
Removed the function export and added all BASe_CLASSES in the MantraData class

Commit:
Removed the function export of data types and added all DATA_TYPES in the Factory class
